### ISSUE TITLE
Is there anyway that I can use C++ library in TEE?

### ISSUE THREAD

**Issue by gutouyu** [OP-TEE/optee_os#2628]
I have some programs wrote in C++ and using as *.so. But I'm trying to run the whole app in TEE. Still don't know if it will work. Or the only way is to re-write the code in C? 

What if I compile C++ library outside using cross-compile tools, then link it to TAs. Is that going to work? (I have seen the QA, and found it is possible to link some c static library. )

Any advice will be appreciated.

**Comment by jenswi-linaro** [comment]
I don't think it will work, exception handling needs special support with isn't there. There's probably more stuff that won't work. I'd recommend re-write the code in C, unless you're prepared to add C++ support the the TA run-time environment.

**Comment by gutouyu** [comment]
@jenswi-linaro  So sad to hear that. :(
I still have some confusions described below:
1. So the exceptions are the main problems, right? If I won't use exceptions, do you think I got a chance?
2. Does out Tee support load dynamic lib(*.so) written in C? Or only support static libs?
3. Can we run C++ code or load C++ library in CA, and then run C code in TA? Is that possible? Or we can only run c code in both CA and TA?

Looking forward to hear from you. Thanks.

**Comment by jenswi-linaro** [comment]
> So the exceptions are the main problems, right? If I won't use exceptions, do you think I got a chance?

Another problem is global initializors and destructors, I don't know if the list stops there or not.

> Does our Tee support load dynamic lib(*.so) written in C? Or only support static libs?

Yes it does, but it has to be special .so files wrapped as a special kind of TA, see https://github.com/OP-TEE/optee_test/tree/master/ta/os_test_lib for an example.

> Can we run C++ code or load C++ library in CA, and then run C code in TA? Is that possible? Or we can only run c code in both CA and TA?

Sure, the client can use C++.
